[1 - 2단계 방탈출 예약 대기] 제프리(홍성호) 미션 제출합니다.

AppleMint98 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?

> https://prolog.techcourse.co.kr/studylogs/4360
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->

규칙3. 모든 원시값과 문자열을 포함한다. 
규칙8. 일급 컬렉션을 쓴다. 

해당 두 규칙을 지키지 못한것 같아요. 

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

안녕하세요 매트 😄 제프리입니다 반갑습니다!

레벨1 출석 미션에서도 코드 리뷰와 피드백으로부터 많은 도움을 받았는데, 레벨2 방탈출 예약 대기 미션에서도 만나게 되었네요. 

[이번 미션의 변경 커밋입니다.](https://github.com/woowacourse/spring-roomescape-waiting/pull/223/files/c93c0c091be75db8539537a7d834f88594ab19b1..e4974526bc4ea34e0a77f4841f9d886b030828ed
)


---

이번 미션에서는 `JdbcTemplate` 으로 작성하던 기존 코드를 `JpaRepository` 로 바꾸는데 초점을 두었습니다!

어떻게 잘 활용할 수 있을지보다는, 일단 JPA를 사용해보자! 가 주된 목표여서 별다른 질문점이 생기지 않았습니다.

이번에도 잘부탁드립니다! 🙇 


### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

Hacanna42 committed:
chore: 이전 코드 가져오기

Hacanna42 committed:
build: JPA 의존성 추가

Hacanna42 committed:
docs: README 추가

Hacanna42 committed:
feat: 기존 도메인 엔티티 매핑

Hacanna42 committed:
refactor: id() -> getId() 변경

Hacanna42 committed:
refactor: 패키지 구조 리팩터링

Hacanna42 committed:
docs: README 업데이트

Hacanna42 committed:
feat: JPA 레포지토리 구현

Hacanna42 committed:
chore: html, js 파일 추가

Hacanna42 committed:
feat: 내 예약 목록 조회 기능 추가

- /reservations-mine API 추가
- js 코드 수정
- 페이지 응답 추가

Hacanna42 committed:
feat: JWT 토큰에 사용자 이름 추가

Hacanna42 committed:
feat: 이메일 중복 체크 기능 추가 및 예외 처리 구현

Hacanna42 committed:
refactor: Exception 처리 리팩터링

Hacanna42 committed:
test: 테스트 코드 리팩터링

Hacanna42 committed:
fix: PK의 GenerationType 수정

Hacanna42 committed:
refactor: 네이티브 쿼리에서 JPQL로 변경

AppleMint98 committed:
refactor: data.sql 제거

hyeonic commented on src/main/java/roomescape/domain/Member.java:
```diff
@@ -0,0 +1,52 @@
+@Getter
+public class Member {
+
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+
+    private Long id;
```
불필요 공백은 제거해주세요!

hyeonic commented on src/main/java/roomescape/domain/Member.java:
```diff
@@ -0,0 +1,52 @@
+        this.password = password;
+    }
+
+    @Override
+    public boolean equals(final Object o) {
+        if (o == null || getClass() != o.getClass()) return false;
+        Member member = (Member) o;
+        return Objects.equals(id, member.id) && Objects.equals(name, member.name) && Objects.equals(role, member.role) && Objects.equals(email, member.email) && Objects.equals(password, member.password);
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, name, role, email, password);
+    }
```
어떠한 목적으로 equals & hashCode를 재정의하신건가요?

applemint98 replied:
지금당장은 엔티티간의 동등성 비교를 해줄 필요가 없지만, 
추후 예외처리나 비즈니스 사용할것 같아 추가하였습니다.

equals 메서드를 재정의할 때 `id != null` 를 추가하였는데, 
id를 할당받기 전 null인 두 객체의 동등성비교가 true가 나오는것을 방지하기 위해 추가하였습니다!

hyeonic replied:
추가로 Entity의 경우 LAZY 등 다양한 기능 제공을 위해 프록시 패턴을 사용하고 있는데요, 어떻게 하면 영속 상태와 아닌 상태를 적절히 구분하여 비교할 수 있는지, 프록시 상태의 객체와 실제 객체에 대한 대한 비교는 어떻게 하면 좋을지도 함께 고려되면 좋을 것 같습니다.

제프리만의 기준이 잘 적립되면 엔티티 전반적으로 통일해봐도 좋을 것 같아요!

applemint98 replied:
감사합니다!!

아직 프록시나, 영속성 컨텍스트에 대해서는 잘 모르지만 
점차 배워가면서 저만의 엔티티 기준을 세우도록 해볼게요 ㅎㅎ

hyeonic commented on src/main/java/roomescape/domain/Member.java:
```diff
@@ -0,0 +1,52 @@
+import roomescape.exception.UnableCreateMemberException;
+
+@Entity
+@NoArgsConstructor
```
기본 생성자가 필요한 이유는 무엇인가요?

applemint98 replied:
JPA에서 엔티티를 변환할 떄 리플렉션을 기반으로 하기 때문에, 
파라미터가 없는 기본 생성자를 먼저 호출 후 DB에서 읽어온 값을 주입해주는 것으로 알고 있습니다.

지우고 실행해보면,

> org.hibernate.InstantiationException: No default constructor for entity 'roomescape.domain.Member'

이런 런타임 에러가 나왔습니다.

AccessLevel을 Protected로 한 이유는, 
private는 리플렉션을 사용할 수 없고, public은 외부에서 빈 생성자로 객체를 생성할 수 있어
외부 생성을 막으면서 JPA를 사용하기 위해 결정했습니다.

hyeonic commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,72 @@
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
`@ManyToOne`는 어떠한 목적으로 활용되나요?

applemint98 replied:
JPA 연관관계 매핑을 위해 사용했습니다.

현재 한 테마나 예약시간에서 여러 예약을 생성할 수 있기에, 
일대다 관계에서 예약을 다 로 잡고 `@ManyToOne` 어노테이션을 추가하였습니다!!

hyeonic replied:
넵 좋습니다! 다만 저는 ManyToOne 같은 애노테이션을 사용할 때 명확하게 어떠한 id를 FK로 사용하였는지 JoinColumn을 통해 명시하는 것을 선호하는데요, 요부분은 취향의 영역일 수 있어서 참고만 부탁드려요.

applemint98 replied:
`JoinColumn` 을 잘 몰랐는데, FK를 명시적으로 표현해줄 수 있다면 표현해주는게 더 좋을것같아요!!
감사합니다 😄

hyeonic commented on src/main/java/roomescape/domain/Reservation.java:
```diff
@@ -0,0 +1,72 @@
+               ", time=" + time +
+               ", theme=" + theme +
+               '}';
+    }
```
어떠한 목적의 toString인가요?

applemint98 replied:
필요없는것 같아요 ㅎㅎ 지우겠습니다! 😄

hyeonic commented on src/test/java/roomescape/domain/MemberTest.java:
```diff
@@ -0,0 +1,26 @@
+//                .isInstanceOf(ReservationException.class)
+//                .hasMessageContaining("예약자명은 2글자에서 10글자까지만 가능합니다.");
+//    }
+//}
```
주석 처리된 테스트는 무엇인가요?

applemint98 replied:
테스트 컴파일이 깨지는 것을 막기위해 주석처리했습니다. 

Service 계층 테스트 보완해서 제출하겠습니다!

hyeonic commented on src/main/java/roomescape/service/MemberService.java:
```diff
@@ -0,0 +1,65 @@
+        Member member = memberRepository.findByEmail(loginMemberParam.email())
+                .orElseThrow(() -> new NotFoundMemberException(loginMemberParam.email() + "에 해당하는 유저가 없습니다."));
+
+        if (!Objects.equals(member.getPassword(), loginMemberParam.password())) {
+            throw new UnauthorizedException("비밀 번호가 일치하지 않습니다.");
+        }
```
member에게 메시지를 던져볼 순 없을까요?

applemint98 replied:
감사합니다!! 객체지향 생활체조 원칙을 놓치고 있었네요... 수정하겠습니다 😄

hyeonic commented on src/main/java/roomescape/service/ReservationService.java:
```diff
@@ -0,0 +1,111 @@
+
+        Reservation savedReservation = reservationRepository.save(reservation);
+        return savedReservation.getId();
+    }
```
이렇게 복잡한 로직이 동반될 때 트랜잭션 처리 없이 원자성을 보장할 수 있을까요?

applemint98 replied:
`@Transactional` 어노테이션을 추가하였습니다!

hyeonic replied:
레벨 2에서는 트랜잭션 관련 개념을 중점적으로 다루는 것은 아니지만 RDB를 쓰는 입장에서 굉장히 중요하다고 생각해요. 이후에 DB 관련하여 학습할 때 JPA에 한정하지 않고 어떻게 하면 일관된 데이터를 읽고 수정할 수 있을지 고민해보면 좋을 것 같습니다!

hyeonic reviewed (changes requested):
안녕하세요 제프리~ 오랜만에 만나뵙게 되어 반갑습니다 ㅎㅎ

간단히 코멘트 남겼으니 확인해주시고 추가로 궁금한 부분 있다면 언제든 편하게 DM 남겨주세요~

AppleMint98 committed:
refactor(domain): 도메인 리팩토링

- 불필요 공백 제거
- equals & hashcode 재정의, id가 null일경우 동등하지 않도록 로직 추가
- 기본생성자 protected 변경
- 불필요한 toString 제거

AppleMint98 committed:
test(domain): 도메인 테스트 추가

- 동등성 판별 로직 테스트 추가
- 유효성 검증 메서드 테스트 추가

AppleMint98 committed:
refactor(service): 도메인 변경에 따른 서비스 수정

AppleMint98 committed:
test(service): 통합테스트 단위테스트로 수정

- SpringBootTest 삭제 후 Mockito 사용

AppleMint98 committed:
refactor(service): 트랜잭션 어노테이션 추가

- 읽기전용 메서드에는 readOnly = true 추가

hyeonic commented on src/main/java/roomescape/exception/global/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,75 @@
+
+@Slf4j
+@ControllerAdvice
+public class GlobalExceptionHandler {
```
coderabbit이 남겨준 것 처럼 핸들링하지 않은 예외를 다루기 위한 핸들러도 추가되면 좋을 것 같습니다!

hyeonic commented on src/main/java/roomescape/persistence/ReservationRepository.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.List;
+
+
+public interface ReservationRepository extends ListCrudRepository<Reservation, Long> {
```
JPA에 대한 이해도가 충분치 않다면 학습 테스트 목적으로 Repository 계층(`@JpaDataTest` 등)에 대한 테스트도 추가해보면 좋을 것 같아요~

applemint98 replied:
이기회에 한번 추가해보겠습니다!! 
감사합니다 매트 😄

hyeonic commented on src/test/java/roomescape/service/MemberServiceTest.java:
```diff
@@ -0,0 +1,100 @@
+import roomescape.service.param.RegisterMemberParam;
+import roomescape.service.result.MemberResult;
+
+@ExtendWith(MockitoExtension.class)
```
대부분 mock을 활용한 테스트가 많아보이네요! mocking을 선호하시는 이유가 있나요??

applemint98 replied:
단위테스트에 더욱 초점을 두기위해 mock을 사용했습니다!

각 클래스별로 단위테스트의 충실도가 높아진다면, 
Service를 테스트하기위해 의존하는 Repository를 모킹해도 충분하다고 생각해요.

추가로 주변 크루들이 Fake 객체를 많이 사용하는 것 같아요. 
저는 Fake는 또다른 유지보수의 대상이 될 뿐더러, (추가적 구현 필요 및 테스트 작성) 
현재의 Repository가 h2메모리 DB이기 때문에 비용이 크지 않다고 생각해
당장은 Fake보다는 mocking을 더 선호합니다!! 😃

hyeonic reviewed (approved):
안녕하세요 제프리! 대부분 요구사항 맞춰서 잘 구현해주신 것 같아서 이후 단계 진행해도 좋을 것 같습니다. 추가적으로 같이 확인해보면 좋을 것 같은 부분 질문 남겨두었으니 같이 고려하여 진행해주세요~

혹여나 제가 놓친 질문이나 추가적으로 궁금한 부분 있다면 언제든 편하게 DM 남겨주세요!
